draft-irtf-cfrg-opaque.md

```diff
@@ -393,152 +391,346 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
+
+Input:
+- seed, pseudo-random byte sequence used as a seed
+
+Output:
+- private_key, a private key
+- public_key, the associated public key
+
+Steps:
+1. private_key = HashToScalar(seed, dst)
+2. public_key = private_key * G
+3. Output (private_key, public_key)
 ~~~
 
-The `custom_identifier` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+HashToScalar(msg, dst): Hash arbitrary input to a scalar in the field of the AKE group as defined in {{I-D.irtf-cfrg-voprf}}, except that dst = "OPAQUE-HashToScalar".
 
 ~~~
-struct {
-  opaque client_private_key[Nsk];
-} SecretCredentials;
+BuildInnerEnvelope(prk, nonce, client_private_key)
 
-struct {
-  opaque server_public_key[Npk];
-  opaque client_identity<0..2^16-1>;
-  opaque server_identity<0..2^16-1>;
-} CleartextCredentials;
+Input:
+- prk, pseudo-random key
+- nonce, a unique nonce of length `Nn`
+- client_private_key, empty value. Not used in this function, it only serves to comply with the API
+
+Output:
+- inner_env, nil value (serves to comply with the API)
+- client_public_key, the client's AKE public key
+
+Steps:
+1. seed = Expand(prk, concat(nonce, "PrivateKey"), Nsk)
+2. _, client_public_key = DeriveAkeKeyPair(seed)
+3. Output (nil, client_public_key)
+~~~
+
+~~~
+RecoverKeys(prk, nonce, inner_env)
+
+Input:
+- prk, pseudo-random key
+- nonce, a unique nonce of length `Nn`
+- inner_env, an InnerEnvelope structure. Not used in this function, it only serves to comply with the API
+
+Output:
+- client_private_key, The encoded client private key for the AKE protocol
+- client_public_key, The encoded client public key for the AKE protocol
+
+Steps:
+1. seed = Expand(prk, concat(nonce, "PrivateKey"), Nh)
+2. client_private_key, client_public_key = DeriveAkeKeyPair(seed)
+4. Output (client_private_key, client_public_key)
 ~~~
 
-These credentials are embedded into the following `Envelope` structure with
-encryption and authentication.
+### External Key mode {#external-mode}
+
+With the external key mode the `EnvelopeMode` value MUST be `external`, and the size `Ne` of the serialized `Envelope` is 1 + Nn + Nm + Nsk.
+
+In this mode, the client's public key is recovered using the private key. An encryption key is generated from the hardened OPRF output and used to encrypt the client's private key, which is then stored encrypted in the `InnerEnvelope`. This allows for applications that prefer to use custom keys.
 
 ~~~
 struct {
-  EnvelopeMode mode;
-  opaque nonce[32];
-  opaque encrypted_creds[Nsk];
+ opaque encrypted_creds[Nsk];
 } InnerEnvelope;
+~~~
 
-struct {
-  InnerEnvelope inner_env;
-  opaque auth_tag[Nm];
-} Envelope;
+encrypted_creds : Encrypted client_private_key. Authentication of this field is ensured with the `AuthTag` in the envelope that covers this `InnerEnvelope`.
+
+If the implementation provides the `client_public_key` to `CreateEnvelope()`, then `BuildInnerEnvelope()` should spare the `RecoverPublicKey()` call as the key is already available.
```

Comment:
```suggestion
If the implementation provides the `client_public_key` to `CreateEnvelope()`, then `BuildInnerEnvelope()` MAY skip the `RecoverPublicKey()` call and use the public key provided.
```

Reply:
`BuildInnerEnvelope()` doesn't need to use the public key